packages/squiggle-lang/src/XYShape.ts

```diff
@@ -482,51 +482,70 @@ export const PointwiseCombination = {
     const t2n = t2.xs.length;
     const outX: number[] = [];
     const outY: number[] = [];
-    let i = -1;
-    let j = -1;
-
-    while (i <= t1n - 1 && j <= t2n - 1) {
-      let x, ya, yb;
-      if ((j === t2n - 1 && i < t1n - 1) || t1.xs[i + 1] < t2.xs[j + 1]) {
-        // if a has to catch up to b, or if b is already done
-        i++;
-
-        x = t1.xs[i];
-        ya = t1.ys[i];
-
-        yb = interpolator(t2, j, x);
-      } else if (
-        (i === t1n - 1 && j < t2n - 1) ||
-        t1.xs[i + 1] > t2.xs[j + 1]
-      ) {
-        // if b has to catch up to a, or if a is already done
-        j++;
-
-        x = t2.xs[j];
-        yb = t2.ys[j];
-
-        ya = interpolator(t1, i, x);
-      } else if (i < t1n - 1 && j < t2n && t1.xs[i + 1] === t2.xs[j + 1]) {
-        // if they happen to be equal, move both ahead
-        i++;
-        j++;
-        x = t1.xs[i];
-        ya = t1.ys[i];
-        yb = t2.ys[j];
-      } else if (i === t1n - 1 && j === t2n - 1) {
-        // finished!
-        i = t1n;
-        j = t2n;
-        continue;
+    // Next index that we want to consume.
+    // Possible range of values: [0..n+2], where:
+    // - `n` signifies "we consumed all points"
+    // - `n+1` signifies "we consumed all points _and_ one more "last x + epsilon" point for extrapolations when shapes don't overlap.
+    // - `n+2` signifies "we consumed all points _and_ "last x + epsilon" point _and_ "first x of another shape - epsilon".
+    // See https://github.com/quantified-uncertainty/squiggle/issues/1400#issuecomment-1666998578 for details why this might be important.
+    let i1 = 0,
+      i2 = 0;
+
+    while (i1 < t1n || i2 < t2n) {
+      let x, y1, y2;
+
+      // First four cases here are for adding two additional points when shapes don't overlap.
+      // TODO - in rare cases, last point of the first shape + epsilon could be exactly first point of the second, which will cause "Xs is not sorted" error later.
+      if (i1 > 0 && i1 === t1n && i2 === 0) {
```

Comment:
I don't like the way I had to code these four conditions, but I couldn't think of anything better that would follow the "each loop iteration adds exactly one outX value" rule.

I could violate that rule, but since `fn` returns a `result`, it would make things complicated and/or slow.

Reply:
This change fixes #1414.

It also _almost_ fixes #1400, but not quite: pointsets are generally not normalized because of precision issues; `symDist -> PointSet.fromDist` is an exception because it lies about its integralSum (I left a comment on that in `SymbolicDist.ts`).

Compare:
https://www.squiggle-language.com/playground#code=eNqrVirOyC8PLs3NTSyqVLIqKSpN1QELuaZkluQXwURSUtMSS3NKnPNTUpWslFIyi0sUbBWCK3P18vKLchNzNAx0DDUV9LSRhQwNDAx0FAw1Y%2FIgAplVqSlATSC9unaZxX5wwZi8zLyS1PSixByENFQA6C6lWgAV9jiw
<img width="1500" alt="Captura de pantalla 2023-08-06 a la(s) 23 22 44" src="https://github.com/quantified-uncertainty/squiggle/assets/89368/b9742d05-4d60-4c9c-8879-dae0a9a9fbaa">

https://squiggle-website-git-pointset-imp-ed491b-quantified-uncertainty.vercel.app/playground#code=eNqrVkpJTUsszSlxzk9JVbJSSsksLlGwVQiuzNXLyy%2FKTczRMNAx1FTQ00YWMjQwMNBRMNSMyYMIZFalpgA1gfTq2mUW%2B8EFY%2FIy80pS04sScxDSUIHg0lwlHaXijPxyICs3sahSyaqkqDS1FgDD8jIe
<img width="1495" alt="Captura de pantalla 2023-08-06 a la(s) 23 23 07" src="https://github.com/quantified-uncertainty/squiggle/assets/89368/60284246-0a87-49ff-837b-f8d6d4d98156">